pkg/recommendedversion/recommended_version.go

```diff
@@ -0,0 +1,268 @@
+}
+
+// sortRecommendedVersionsDescending will convert the array of recommended
+// versions into an array sorted in descending order of semver
+func sortRecommendedVersionsDescending(recommendedVersions []string) ([]string, error) {
+	// Trim any spaces around the version strings and remove duplicates
+	finalVersions := make([]string, 0, len(recommendedVersions))
+	alreadyPresent := make(map[string]bool)
+	for _, newVersion := range recommendedVersions {
+		trimmedVersion := strings.TrimSpace(newVersion)
+		if trimmedVersion != "" && !alreadyPresent[trimmedVersion] {
+			finalVersions = append(finalVersions, trimmedVersion)
+			alreadyPresent[trimmedVersion] = true
+		}
+	}
+
+	// Now sort the versions, then reverse the order
+	err := utils.SortVersions(finalVersions)
+	if err != nil {
+		return nil, err
+	}
+
+	// Reverse the order so it is descending
+	for i := len(finalVersions)/2 - 1; i >= 0; i-- {
+		opp := len(finalVersions) - 1 - i
+		finalVersions[i], finalVersions[opp] = finalVersions[opp], finalVersions[i]
+	}
+	return finalVersions, err
+}
+
+func getRecommendationDelayInSeconds() int {
+	delay := recommendedVersionCheckDelaySeconds
+	delayOverride := os.Getenv(constants.ConfigVariableRecommendVersionDelayDays)
+	if delayOverride != "" {
+		delayOverrideValue, err := strconv.Atoi(delayOverride)
+		if err == nil {
+			if delayOverrideValue >= 0 {
+				// Convert from days to seconds
+				delay = delayOverrideValue * 24 * 60 * 60
+			} else {
+				// When the configured delay is negative, it means the value
+				// should be in seconds.  This is used for testing purposes.
+				delay = -delayOverrideValue
+			}
+		}
+	}
+	return delay
+}
+
+func shouldCheckVersion() bool {
+	delay := getRecommendationDelayInSeconds()
+	if delay == 0 {
+		// The user has disabled the version check
+		return false
+	}
+
+	// Get the last time the version check was done
+	var lastCheck time.Time
+	err := datastore.GetDataStoreValue(dataStoreLastVersionCheckKey, &lastCheck)
+	if err != nil {
+		return true
+	}
+
+	return time.Since(lastCheck) > time.Duration(delay)*time.Second
+}
+
+func printVersionRecommendations(writer io.Writer, currentVersion, _, minor, patch string) {
+	// Only print the message for minor and patch versions.
+	// We don't print anything for major versions because they are breaking changes
+	// and it will take a while for the user to upgrade to a new major version.
+	if (minor == "" || !utils.IsNewVersion(minor, currentVersion)) &&
+		(patch == "" || !utils.IsNewVersion(patch, currentVersion)) {
+		// The current version is the best recommended version
+		return
+	}
+
+	// Put a delimiter before this notification so the user
+	// can see it is not part of the command output
+	fmt.Fprintln(writer, "\n==")
+	fmt.Fprintf(writer, "Note: A new version of the Tanzu CLI is available. You are at version: %s.\n", currentVersion)
+	fmt.Fprintln(writer, "To benefit from the latest security and features, please update to a recommended version:")
```

Comment:
(my previous comment was gone, re-adding in case you missed it)

nit: "latest security and features" doesn't read right. You mean "latest security improvements and features", or "latest security and feature enhancements", perhaps?

That said, is that statement always true with every update?
I mean, is it possible a patch does not contain security-related enhancements at all?
...
which leads me to thinking (as a future feature enhancement) if the Central Config for recommended versions should also contain an optional one-liner or two describing the main purpose of the update. 
That way, the information can be used here to augment an otherwise generic message about the recommended version.

Reply:
Now that we won't recommend a new major version, there will only be two lines at most.
I think @anujc25 is right, users will be able to understand the difference, so I have removed the explanations since they were contentious.

But I do like your suggestion of adding an "optional one-liner or two describing the main purpose of the update".
This makes me worry about the future-proofing of the current recommendedVersions format.  If we want to add a description later, we'll need to use a new key since the current key/value is read as an array of strings, which is not cleanly extensible.

We could instead read it as a struct containing a `[]string`, so that we could add a new field later.

What do you think?

Reply:
After discussion we agreed to try to do this.
It turned out to be easy enough that I added it as a top-commit on this PR.